noir-projects/noir-protocol-circuits/crates/types/src/merkle_tree/variable_merkle_tree.nr

```diff
@@ -16,6 +16,35 @@ unconstrained fn get_height(input: u32, start: u8) -> u8 {
     height
 }
 
+/// Calculates the previous power of 2.
+///
+/// The previous power of 2 is the largest power of 2 that is less than or equal to a number.
+/// For example:
+/// - For 7, previous power of 2 is 4 (2^2)
+/// - For 10, previous power of 2 is 8 (2^3)
+/// - For 16, previous power of 2 is 16 (2^4)
+///
+/// # Arguments
+/// * value - Integer value for which we need to find the previous power of 2
+///
+/// # Returns
+/// The previous power of 2
+fn get_prev_power_2(value: u32) -> u32 {
+    // Find size of tree required
+    let height = unsafe {
+        //@safety This is safe because we derive next and previous powers of 2 below based on the height
+        // and the powers are then checked against the input value.
+        get_height(value, 0)
+    };
+
+    let next_power_2 = 2 << height;
+    let prev_power_2 = next_power_2 / 2;
+    assert((value == 0) | (value == 1) | (value > prev_power_2));
+    assert(value <= next_power_2);
+
+    prev_power_2
+}
```

Comment:
As proposed by Nico in a PR down the stack. Made sense to separate this into a separate func as it makes it easier to figure out how the height should be constrained.

Reply:
Moved to a separate PR https://github.com/AztecProtocol/aztec-packages/pull/11165